cores/arduino/stm32/bootloader.c

```diff
@@ -2,7 +2,144 @@
 
 #include "stm32_def.h"
 #include "backup.h"
+#include "stm32yyxx_ll_system.h"
+#include "usbd_if.h"
 
+/*
+ * STM32 built-in bootloader in system memory support
+ */
+/* Private definitions to manage system memory address */
+#define SYSMEM_ADDR_COMMON 0xFFF
+
+typedef struct {
+  uint32_t devID;
+  uint32_t sysMemAddr;
+} devSysMemAddr_str;
+
+devSysMemAddr_str devSysMemAddr[] = {
+#ifdef STM32F0xx
+  {0x440, 0x1FFFEC00},
+  {0x444, 0x1FFFEC00},
+  {0x442, 0x1FFFD800},
+  {0x445, 0x1FFFC400},
+  {0x448, 0x1FFFC800},
+  {0x442, 0x1FFFD800},
+#elif STM32F1xx
+  {0x412, 0x1FFFF000},
+  {0x410, 0x1FFFF000},
+  {0x414, 0x1FFFF000},
+  {0x420, 0x1FFFF000},
+  {0x428, 0x1FFFF000},
+  {0x418, 0x1FFFB000},
+  {0x430, 0x1FFFE000},
+#elif STM32F2xx
+  {0x411, 0x1FFF0000},
+#elif STM32F3xx
+  {SYSMEM_ADDR_COMMON, 0x1FFFD800},
+#elif STM32F4xx
+  {SYSMEM_ADDR_COMMON, 0x1FFF0000},
+#elif STM32F7xx
+  {SYSMEM_ADDR_COMMON, 0x1FF00000},
+#elif STM32G0xx
+  {SYSMEM_ADDR_COMMON, 0x1FFF0000},
+#elif STM32G4xx
+  {SYSMEM_ADDR_COMMON, 0x1FFF0000},
+#elif STM32H7xx
+  {0x450, 0x1FF00000},
+#elif STM32L0xx
+  {SYSMEM_ADDR_COMMON, 0x1FF00000},
+#elif STM32L1xx
+  {SYSMEM_ADDR_COMMON, 0x1FF00000},
+#elif STM32L4xx
+  {SYSMEM_ADDR_COMMON, 0x1FFF0000},
+#elif STM32WBxx
+  {SYSMEM_ADDR_COMMON, 0x1FFF0000},
+#else
+#warning "No system memory address for this serie!"
+#endif
+  {0x0000, 0x00000000}
+};
+
+uint32_t getSysMemAddr(void)
+{
+  uint32_t sysMemAddr = 0;
+  if (devSysMemAddr[0].devID == SYSMEM_ADDR_COMMON) {
+    sysMemAddr = devSysMemAddr[0].sysMemAddr;
+  } else {
+    uint32_t devId = LL_DBGMCU_GetDeviceID();
+    for (uint32_t id = 0; devSysMemAddr[id].devID != 0; id++) {
+      if (devSysMemAddr[id].devID == devId) {
+        sysMemAddr = devSysMemAddr[id].sysMemAddr;
+        break;
+      }
+    }
+  }
+  return sysMemAddr;
+}
+
+/* Request to jump to system memory boot */
+WEAK void jumpToBootloaderRequested(void)
+{
+  enableBackupDomain();
+  setBackupRegister(SYSBL_MAGIC_NUMBER_BKP_INDEX, SYSBL_MAGIC_NUMBER_BKP_VALUE);
+  NVIC_SystemReset();
+}
+
+/* Jump to system memory boot from user application */
+WEAK void jumpToBootloader(void)
+{
+  enableBackupDomain();
+  if (getBackupRegister(SYSBL_MAGIC_NUMBER_BKP_INDEX) == SYSBL_MAGIC_NUMBER_BKP_VALUE) {
+    setBackupRegister(SYSBL_MAGIC_NUMBER_BKP_INDEX, 0);
+
+#ifdef USBCON
+    USBD_reenumerate();
+#endif
+    void (*sysMemBootJump)(void);
+
+    /**
+     * Get system memory address
+     *
+     * Available in AN2606 document:
+     * Table xxx Bootloader device-dependent parameters
```

Comment:
Maybe this reference to the AN should be moved up to the definition of the table? That would have helped me understand that code and its source better?

Reply:
Right